datalad/runner/nonasyncrunner.py

```diff
@@ -35,226 +48,499 @@
+            `stdout`, `stderr`, and `shell` will be overwritten internally.
         """
-        super().__init__(daemon=True)
-        self.stdin_data = stdin_data
-        self.process = process
-        self.stdin_fileno = stdin_fileno
-        self.queue = q
-        self.command = command
 
-    def __str__(self):
-        return (
-            f"WriterThread(stdin_data[0 ... {len(self.stdin_data) - 1}], "
-            f"{self.process}, {self.stdin_fileno}, {self.command})")
-
-    def run(self):
-        lgr.log(5, "%s started", self)
+        self.cmd = cmd
+        self.protocol_class = protocol_class
+        self.stdin = stdin
+        self.protocol_kwargs = protocol_kwargs or {}
+        self.timeout = timeout
+        self.popen_kwargs = popen_kwargs
+        self.catch_stdout = self.protocol_class.proc_out is not None
+        self.catch_stderr = self.protocol_class.proc_err is not None
+        self.generator = self.protocol_class.generator is not None
+
+        self.write_stdin = False
+        self.stdin_queue = None
+        self.protocol = None
+        self.process = None
+        self.process_stdin_fileno = None
+        self.process_stdout_fileno = None
+        self.process_stderr_fileno = None
+        self.stderr_reader_thread = None
+        self.stderr_enqueueing_thread = None
+        self.stdout_reader_thread = None
+        self.stdout_enqueueing_thread = None
+        self.stdin_writer_thread = None
+        self.stdin_enqueueing_thread = None
+        self.process_running = False
+        self.fileno_mapping = None
+        self.fileno_to_file = None
+        self.output_queue = None
+        self.active_file_numbers = None
+
+    def run(self) -> Union[Any, Generator]:
+        """
+        Run the command as specified in __init__.
+
+        Returns
+        -------
+        Any
+            If the protocol is not a subclass of `GeneratorMixIn`, the
+            result of protocol._prepare_result will be returned.
+
+        Generator
+            If the protocol is a subclass of `GeneratorMixIn`, a Generator
+            will be returned. This allows to use this method in constructs
+            like:
+
+                for protocol_output in runner.run():
+                    ...
+
+            Where the iterator yields whatever protocol.pipe_data_received
+            sends into the generator.
+            If all output was yielded and the process has terminated, the
+            generator will raise StopIteration(return_code), where
+            return_code is the return code of the process. The return code
+            of the process will also be stored in the "return_code"-attribute
+            of the runner. So you could write:
+
+               gen = runner.run()
+               for file_descriptor, data in gen:
+                   ...
+
+               # get the return code of the process
+               result = gen.return_code
+        """
+        if isinstance(self.stdin, (int, IO, type(None))):
```

Comment:
May be more straightforward and explicit

```suggestion
        if self.stdin is None or isinstance(self.stdin, (int, IO)):
```
?

Reply:
If actually find the original version more straightforward, but I am not dogmatic about that. Feel free to commit.